### PR TITLE
Guard against null HttpContext

### LINKED ISSUE #6787: `WrappedFutureCallback` should handle `null` `HttpContext` objects
The `WrappedFutureCallback.getResponse(HttpContext)` method does not handle a `null` input, and `WrappedFutureCallback.completed(T)` does not check if `httpContext` is `null` before calling `getResponse(httpContext)`. 

According to the [async client docs](https://hc.apache.org/httpcomponents-client-5.1.x/current/httpclient5/apidocs/org/apache/hc/client5/http/impl/async/CloseableHttpAsyncClient.html#execute(org.apache.hc.core5.http.nio.AsyncRequestProducer,%20org.apache.hc.core5.http.nio.AsyncResponseConsumer,%20org.apache.hc.core5.http.nio.HandlerFactory,%20org.apache.hc.core5.http.protocol.HttpContext,%20org.apache.hc.core5.concurrent.FutureCallback)), the `HttpContext` is optional and can be `null`. It seems like this should be handled in the `WrappedFutureCallback` implementation.

It looks like the `completed()`, `failed()` and `cancelled()` methods all make the same assumptions, so presumably they'd all throw a `NullPointerException`, although the case I was running into was happening in `completed()`.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/b0738a54628fe52c5e7bea5e136dada0865cf674/instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java#L185

### PR DESCRIPTION
**Pull request by breedx-splk**
Resolves #6787

### CODE REVIEW

**Comment by breedx-splk** [comment]
#6791 should address the test failure.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java [hunk 1]
```diff
@@ -159,7 +160,7 @@ public void sendRequest(HttpRequest request, EntityDetails entityDetails, HttpCo
     private static final Logger logger = Logger.getLogger(WrappedFutureCallback.class.getName());
 
     private final Context parentContext;
-    private final HttpContext httpContext;
+    @Nullable private final HttpContext httpContext;
```
Comment (trask) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java:163]: nullable on the field in the other class too?

**Review by trask — APPROVED** [review]

File: instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java (hunk 1 above)
Comment (breedx-splk) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java:163]: Doh! yeah...will do.

**Review by laurit — COMMENTED** [review]
I don't think that this is the best approach. Considering that if the passed `HttpContext` is null a blank one will be created a few calls deeper we might as well create it here. This would avoid loosing the telemetry that requires access to response.

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by breedx-splk** [comment]
> I don't think that this is the best approach. 

I will never assert that any change I have ever or will ever make will be the _"best"_.

>Considering that if the passed `HttpContext` is null a blank one will be created a few calls deeper we might as well create it here. This would avoid loosing the telemetry that requires access to response.

I'm not following this. Where does a blank `HttpContext` get created a few calls deeper? And what kind of response attributes are even useful when the context is null? Isn't it all worthless without context? Where does a null response inhibit other telemetry?

**Comment by laurit** [comment]
> > Considering that if the passed `HttpContext` is null a blank one will be created a few calls deeper we might as well create it here. This would avoid loosing the telemetry that requires access to response.
> 
> I'm not following this. Where does a blank `HttpContext` get created a few calls deeper? And what kind of response attributes are even useful when the context is null? Isn't it all worthless without context? Where does a null response inhibit other telemetry?

See https://github.com/apache/httpcomponents-client/blob/7ab435c271b6ff3bde4e737c2b02ba873ffa8b55/httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/InternalAbstractHttpAsyncClient.java#L177 for where http client handles null context.

Try adding null context to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/da3eecfd9f88e53abf0cc63308e03f2a90d90299/instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientTest.java#L91 and https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/da3eecfd9f88e53abf0cc63308e03f2a90d90299/instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientTest.java#L96 when you run the test with your modifications you'll see that some tests fail because `http.status_code` and `http.flavor` are not filled because they are extracted from the response. Change https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/da3eecfd9f88e53abf0cc63308e03f2a90d90299/instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java#L70 to `@Advice.Argument(value = 3, readOnly = false) HttpContext httpContext` and add 
```
      if (httpContext == null) {
        httpContext = new BasicHttpContext();
      }
```

**Comment by breedx-splk** [comment]
Ok, I think I see where you're coming from. The fact that the client defaults the null `httpContext` to a `BasicHttpContext` inside of `doExecute()` is lost on us, because our instrumentation is sitting above that. Alright, I changed the test code to explicitly pass a null context and tried 3 things:

1. No null guard in `WrappedFutureCallback`, no defaulting `httpContext` in`ClientAdvice`. 
Result: test hangs (presumably due to NPE)
2. Add null guard in `WrappedFutureCallback`, no defaulting `httpContext` in `ClientAdvice`. 
Result: Test failures due to missing http flavor and status code
3. Add null guard and default `httpContext` in `ClientAdvice`. 
Result: Test failures due to number of expected spans being wrong

It's feeling like one of the changes (either test code or instrumentation) is a semantic change. I'm not sure which it is....but explicitly passing in a null context caused the tests to fail before this change and with this change.

If you feel strongly about the null guard being a bad solution, I can close this.

**Comment by trask** [comment]
> 3. Add null guard in `WrappedFutureCallback`, and default `httpContext` in `ClientAdvice`.
>    Result: Test failures due to number of expected spans being wrong

@breedx-splk did you maybe miss making the `httpContext` writable in `ClientAdvice`?

```
@Advice.Argument(value = 3, readOnly = false) HttpContext httpContext
```

**Comment by breedx-splk** [comment]
> @breedx-splk did you maybe miss making the `httpContext` writable in `ClientAdvice`?

🤦🏻 yes...that was it. Thanks. I'll push a change.

**Comment by trask** [comment]
@breedx-splk can you rebase? github UI appears to be lying about your change set (at least to me), I've seen this a couple of times recently and rebasing to latest `main` should hopefully clean it up

File: instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java [hunk 2]
```diff
@@ -159,7 +164,7 @@ public void sendRequest(HttpRequest request, EntityDetails entityDetails, HttpCo
...
```
Comment (trask) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpAsyncClientInstrumentation.java:167]:
```suggestion
    private final HttpContext httpContext;
```

**Review by trask — APPROVED** [review]

**Pull request merged**